src/coreclr/debug/ee/controller.cpp

```diff
@@ -6410,7 +6410,8 @@ void DebuggerStepper::TrapStepOut(ControllerStackInfo *info, bool fForceTraditio
                     continue;
 
                 dji = info->m_activeFrame.GetJitInfoFromFrame();
-                _ASSERTE(dji != NULL);
+                if (dji == NULL)
```

Comment:
How do we end up with a null DJI? They should be there on JitComplete

Reply:
That occasionally occurs and is the reason for so many of the asserts. One of the reasons is the issue mentioned above about missing Tier1 data on debug attach. Also, this can return `NULL` if an exception is thrown during the `GetJitInfoFromFrame()` call.

Additionally, handling a `NULL` is done by most callers of this API. The one above and one other are the only ones that prohibit `NULL` but as far as I can see this is possible and at the moment, very common.

Reply:
I assume that the other callsites which allow NULL DJIs are doing the lookup on a broader set of inputs. In this case we already confirmed at 6398 and 6406 that the frame we are looking up is a Frameless managed method which (I think) is synonymous with "this method is implemented by R2R or jitted code". If DJI is NULL here it probably means there is a bug in the code that is looking up the DJI. In the case where this assert fires do you know what the code at the IP is? The IP is in info->m_activeFrame->registers.

Reply:
by 'bug in lookup' I am including issues where the lookup was expected to populate an entry into the cache just in time. If that IP is pointing to the tier1 code and the lookup doesn't return a DJI for that code I considering that a bug in lookup. Can you find out what the IP represents?

Reply:
Ah. Yes, the IP is in the Tier1 portion. I can see that in the logs and it states the following. I've updates the logs slightly.

The `0x00007FFA3CDAAE30` is an IP in Tier1 and `0x00000195DC6C18D0` is the initial DJI. There is a single one and I've confirmed this is the ReadyToRun entry not the Tier1.

```
TID 1bb2c: DJI:GJIBA finding DJI corresponding to addr 0x00007FFA3CDAAE30, starting with 0x00000195DC6C18D0
TID 1bb2c: CRI::GCRI: simple case: CodeRegionInfo* 0x00000195DC6C1908
TID 1bb2c: DJI:GJIBA: pbAddr 0x00007FFA3CDAAE30 is not in code 0x00007FFA9FE5E570 (size:0x0000000000000269)
TID 1bb2c: DJI:GJIBA couldn't find a DJI corresponding to addr 0x00007FFA3CDAAE30
TID 1bb2c: Couldn't find a DJI by address 0x00007FFA3CDAAE30, so it might be a stub or thunk
TID 1bb2c: StubManager::TraceStub: addr 0x00007FFA3CDAAE30 is managed code
TID 1bb2c: Address not thunked - must be to unJITted method, or normal managed method lacking a DJI!
```

Reply:
I think I found the issue.

[See https://github.com/dotnet/runtime/blob/main/src/coreclr/debug/ee/debugger.cpp#L2700-L2720](https://github.com/dotnet/runtime/blob/f99ba2e2b8fbf03be5058ad23e8cdce9c4a09da6/src/coreclr/debug/ee/debugger.cpp#L2700-L2720)

Reply:
I need to head out now but based on the logs and the comments about not handling code versioning properly I believe this is an issue at the very least. It is likely never hit outside of attach scenarios.

Reply:
Yeah, that looks broken. If you wanted to fix that in this PR I think it would be a combo of g_pEEInterface->GetNativeCodeStartAddress(addr) and pDmi->FindOrCreateInitAndAddJitInfo().

Fwiw, a bunch of that stuff in GetJitInfoWorker looks unnecessarily complicated. I think logic like this would work fine:
```
...

PCODE startAddr = g_pEEInterface->GetNativeCodeStartAddress(pbAddr);
if(startAddr == NULL)
{
    // if the address wasn't in jitted code we'll also check to see if it is a stub that leads to jitted code
    TraceDestination trace;
    g_pEEInterface->TraceStub(pbAddr, &trace);
    if(trace.GetTraceType() == TRACE_MANAGED && pbAddr != trace.GetAddress())
    {
        startAddr = trace.GetAddress();
   }
}
if(startAddr != NULL)
{
    dji = dmi->FindOrCreateInitAndAddJitInfo(fd, startAddr);
}
```

Reply:
Now that you fixed the code in GetJitInfoWorker() is there a known case where we expect this callsite to get back a NULL by design? I know it is possible in general for GetJitInfo() to return NULL, but my understanding is that this code path is constrained enough that it doesn't hit those cases.